IC-1462 extend the error space for update requests (#143)
dfinity/interface-spec: spec/index.adoc

* `method_name` (`text`): Name of the canister method to call
* `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:
* 202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use <<http-read-state,`read_state`>> to determine the status of the call.
* 200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the user error. This return type

> is used when the error can be handled by the canister owner (e.g. when a canister is out of cycles) 

Reviewer: The previous sentence talks about "user" error (sounds like it's triggered by the request sent by the user) while the last sentence talks about "canister owner" to fix the error (although canister running out of cycles needn't be only fixed by the controller). Could you please clarify what types of error are expected there? @rumenov

Author: I am not an expert [on execution errors](https://github.com/dfinity/ic/blob/master/rs/http_endpoints/public/src/common.rs#L80) but I believe most of the issues here should be returned as OK 200

@dsarlis any suggestions how to improve the wording here ?

Reviewer 2: If we do this, wouldn't then it mean that 4xx and 5xx can be returned only when the request is malformed (or something along those lines) or the replica is unavailable respectively? And therefore, any IC specific error would be returned via 200 with non-empty body containing the details of the error?

Reviewer 2: Oh wait, too early in the morning, I misread the bullet points. I think we can simply say something like:
```
200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the IC specific error encountered.
```
and completely drop the last part. Some of the errors could be handled only by the canister owner or they could be handled by the client too, so trying to be more specific doesn't help much.

Reviewer 2: Also, I think we should specify the body in the case where an error is returned. I'd expect it would be quite similar to the error case for a query call. So, something like:

```
The body is a CBOR map with the following fields:
* `status` (`text`): `rejected`
* `reject_code` (`nat`): The reject code (see <<reject-codes>>).
* `reject_message` (`text`): a textual diagnostic message.
* `error_code` (text): an optional implementation-specific textual error code (see <<error-codes>>).
```
(maybe we can skip `status` in this case though as it would only be non-empty in the error case anyway).

Author: @dsarlis can you please take another look ?

Reviewer 2: Lgtm!

* `method_name` (`text`): Name of the canister method to call
* `arg` (`blob`): Argument to pass to the canister method

The HTTP response to this request can have the following responses:

* 202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use <<http-read-state,`read_state`>> to determine the status of the call.

> * 200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the IC specific error encountered. The body is a CBOR map with the following fields:

Reviewer 3: I'm not a fan of HTTP 2xx indicating an error. 2xx generally indicate success - see also [here](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).

Author: It was success in the sense the request reached the execution environment. Compared to all alternatives this is maybe the least confusing we found given we want to distinguish between canister/execution errors and IC platform errors.

Reviewer 3: `execution pre-processing error` is this something like failed to decode candid or rejected by guard method? 

Anything above 5xx has no canonical meaning, why not just use 7xx for it? If I as a dev see 200, I am sure everything worked out fine, which it obviously didn't

Reviewer 3: I just saw `All valid status codes are within the range of 100 to 599, inclusive.` in the RFC... blegh.

Looking through the RFC, I feel like for a pre-processing error a lot would be more appropriate than 200. The first sentence of [200](https://www.rfc-editor.org/rfc/rfc9110.html#name-200-ok) is `The 200 (OK) status code indicates that the request has succeeded.`, which is in direct conflict to a pre-processing **error** IMO. [202](https://www.rfc-editor.org/rfc/rfc9110.html#name-202-accepted) feels better, but still it says `accepted for processing, but the processing has not been completed` which is partially true, but processing _has_ completed, with an error. Some codes in the 4xx range seem more valid on first glance (e.g. [406 Not acceptable](https://www.rfc-editor.org/rfc/rfc9110.html#name-406-not-acceptable) or [412 Precondition failed](https://www.rfc-editor.org/rfc/rfc9110.html#name-412-precondition-failed)), but they are talking about headers, which also seems wrong for this case. Maybe [422 Unprocessable content](https://www.rfc-editor.org/rfc/rfc9110.html#name-422-unprocessable-content) would be valid - it talks about `unable to process the contained instructions`.

Author: > If I as a dev see 200, I am sure everything worked out fine, which it obviously didn't

Even today, the spec explicitly says that receiving 200 or even 202 means nothing. You should pool the state to make sure your update call made it through. Given the async nature of the protocol I don't think we have much choice here.

Reviewer 4: I agree with Severin that 2xx codes should not be used for platform errors. Aside from developer confusion, it also reduces compatibility with existing (and future) HTTP aware tooling that understands HTTP codes. Differentiating between platform errors and canister errors using the body fields sounds better, in my opinion.

Reviewer 5: There is a difference to be made between the IC/subnet/replica (the application, if you want) and the HTTP API used to access it. The HTTP status codes (whether 2xx, 4xx or 5xx) only refer to the API, which acts as a proxy to the application. To use the envelope and content analogy at the top of the *Call* section, the status codes refer to the delivery of the envelope.

The alternative (and what we have now, for the most part) is to try and make the HTTP API entirely transparent. But that means either:
* Using the same status code to report on both API and IC errors. E.g. using 404 for "no such API endpoint", "no such canister" and "no such canister method". Similarly, and more confusingly, returning 5xx errors both when the API encounters a server error; when there's an internal replica/sandbox error; and when the canister logic produces an error or panics.
* Or designating some made-up 4xx and 5xx status codes to reflect replica or canister errors. At which point (i) they are not immediately understandable to an HTTP client; and (ii) they are not all that different from returning a 2xx status and a body clarifying the error. Furthermore, a 200 or 202 response to a `call` endpoint by no means guarantees that the subnet will ever induct the ingress message into a block. I.e. we already have the case where a replica replies with 200 or 202, only for the ingress message to disappear without a trace.

Maybe part of the confusion also stems from explicitly stating that 202 stands for success and 200 for error. It would probably be cleaner and clearer to say that a 2xx response means that the API delivered the envelope to the replica and the outcome, if any, is in the body of the response.

Reviewer 3: > Even today, the spec explicitly says that receiving 200 or even 202 means nothing. [...] Given the async nature of the protocol I don't think we have much choice here.

Fair enough. I still don't like how 2xx means nothing, but I guess then we have found the least bad option.

> * 200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the IC specific error encountered. The body is a CBOR map with the following fields:

Reviewer: Could we please list the cases explicitly? Currently, the reference implementation and its test suite does not handle this so I would like to know what to change there?

@rumenov

Author: I don't think this is reasonable. The [list is ever](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/http_endpoints/public/src/common.rs?L80) growing.

Author: > I see. Would they all become 200? Then it could make sense to list which 4xx/5xx are not transient.

I believe they all become 200. @dsarlis to confirm

Reviewer 2: All these errors would become 200 (plus the cbor body that has the IC error details).

> Then it could make sense to list which 4xx/5xx are not transient.

I'm not sure I understood this part.

Reviewer: > Then it could make sense to list which 4xx/5xx are not transient.

Given that all IC-specific errors would become 200, my comment is irrelevant (I wanted to list IC-specific errors - like canister not found - that still return 4xx/5xx instead of 200).

** `reject_code` (`nat`): The reject code (see <<reject-codes>>).
** `reject_message` (`text`): a textual diagnostic message.

> ** `error_code` (text): an optional implementation-specific textual error code (see <<error-codes>>).

Reviewer: ```suggestion
** `error_code` (`text`): an optional implementation-specific textual error code (see <<error-codes>>).
```

> * 4xx HTTP status when the is error caused by the client.

Reviewer: ```suggestion
* 4xx HTTP status when the error is caused by the client.
```

> * 4xx HTTP status for client errors (e.g. malformed request or canister not found). Retrying the request will likely have the same outcome.

Reviewer: > Retrying the request will likely have the same outcome.

I see http status codes like 429 Too Many Requests in the overview of http status codes which should be possible to retry. So I wonder we could rephrase this sentence, e.g., by listing the status codes that do not make sense to retry?

@rumenov

Author: I don't think it makes sense. Then we will be repeating the HTTP protocol standard.

Reviewer: Sure, I don't want to repeat the HTTP protocol standard. Instead I want to explicitly list status codes produced by the IC due to IC-specific failures (canister not found is an example that likely has the same outcome unless the canister is created in the meantime as thus should be listed as a case to which the sentence "Retrying the request will likely have the same outcome" applies).

Author: Given this - https://github.com/dfinity/interface-spec/pull/143#discussion_r1151721927, there won't be any execution errors that map to 4xx/5xx errors.

Reviewer: Perfect! I've made a suggestion to drop "canister not found" and refine the sentence on retrying.

> * 4xx HTTP status for client errors (e.g. malformed request or canister not found). Retrying the request will likely have the same outcome.

Reviewer 5: The suggestion seems too generic: "doing X might fail".

If anything, I'd rather mention 429 explicitly. Or say something along the lines of "The client should not retry the request without modification" (in some cases, said modification would be to e.g. use a different URL or authenticate; in other cases, it would simply mean "wait until the conditions are right").

> * 4xx HTTP status for client errors (e.g. malformed request or canister not found). Retrying the request will likely have the same outcome.

Reviewer: I mentioned 429 explicitly now. I did not go for the sentence "The client should not retry the request without modification" because with 429 you can make the very same request again.

> * 4xx HTTP status when the error is caused by the client.
> * 5xx HTTP status when the server is aware that it has encountered an error or is otherwise incapable of performing the request. 

Reviewer 5: Personal opinion (feel free to ignore): I'dd add some examples and some clarifications for the 4xx and 5xx cases. Something along the lines of (although I'm not sure the examples are entirely accurate):

```suggestion
* 4xx HTTP status for client errors (e.g. malformed request or canister not found). Retrying the request will likely have the same outcome.
* 5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.
```

> * 5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.

Reviewer 3: Would this include e.g. `not enough cycles`? In this case, I'd like to see specific documented cases that can be 
matched against, similar to the above `error_code` because  some errors can be treated in some way by the developer.

On second thought, why not also allow this for any non-200 error code? There will always be use cases for differentiating between different error conditions, and only using HTTP status codes feels too limiting

Author: `not enough cycles` would be 2xx error

> On second thought, why not also allow this for any non-200 error code? There will always be use cases for differentiating between different error conditions, and only using HTTP status codes feels too limiting

Not sure what you exactly mean by this .

Reviewer 3: Some of the 4xx or 5xx status codes have multiple root causes, some of which are recoverable for the client, but at the moment they can only be differentiated with the error message, which may change at any point in time. If we had an error code for these too, I'd be much more comfortable to implement automatic recovery processes.

Author: ideally figuring out if it is recoverable or not the client should use only the status code of the 4xx or 5xx error and not the body. This is part of the HTTP spec to.

Reviewer 3: Then can we please guarantee that one cause of error = one status code? Otherwise it's back to matching on the body since there are no guarantees - having an `error_code` field would make this a lot easier.

One use case I have in mind: a non-controller asks for the canister status. I want to show the user an explanation of controllers / nice error message. On the replica, I can match on HTTP 403, but on ic-ref, it returns 400, where I have to match on the body for "wrong sender" because anything could return a 400.

Similarly, AFAIK 503 only happens if a query happens to hit a replica that is behind, therefore I should retry my request, but who knows? It could be a problem where retrying does not help
